src/Compilers/CSharp/Portable/Lowering/DiagnosticsPass_ExpressionTrees.cs

```diff
@@ -1045,5 +1045,32 @@ public override BoundNode VisitCollectionExpression(BoundCollectionExpression no
 
             return base.VisitCollectionExpression(node);
         }
+
+        public override BoundNode VisitIfStatement(BoundIfStatement node)
```

Comment:
>VisitIfStatement [](http://example.com/codeflow?start=34&length=16)

Consider deriving `DiagnosticsPass` from `BoundTreeWalkerWithStackGuardWithoutRecursionOnTheLeftOfBinaryOperator`. That should allow removing this override and simplifying `VisitBinaryOperator`.

Reply:
> Consider deriving `DiagnosticsPass` from `BoundTreeWalkerWithStackGuardWithoutRecursionOnTheLeftOfBinaryOperator`. That should allow removing this override and simplifying `VisitBinaryOperator`.

I'll open a follow-up issue for this refactoring. There is another PR coming about binary patterns. Will take care of all of them together.

Reply:
Opened #75319.